internal/statemachine/classic_states.go

```diff
@@ -296,6 +311,101 @@ func (stateMachine *StateMachine) createChroot() error {
 	return nil
 }
 
+// add PPAs to the apt sources list
+func (stateMachine *StateMachine) addExtraPPAs() error {
+	var classicStateMachine *ClassicStateMachine
+	classicStateMachine = stateMachine.parent.(*ClassicStateMachine)
+
+	// create /etc/apt/sources.list.d in the chroot if it doesn't already exist
+	sourcesListD := filepath.Join(stateMachine.tempDirs.chroot, "etc", "apt", "sources.list.d")
+	err := osMkdir(sourcesListD, 0755)
+	if err != nil && !os.IsExist(err) {
+		return fmt.Errorf("Failed to create apt sources.list.d: %s", err.Error())
+	}
+
+	// now create the ppa sources.list files
+	tmpGPGDir, err := osMkdirTemp("/tmp", "ubuntu-image-gpg")
+	if err != nil {
+		return fmt.Errorf("Error creating temp dir for gpg imports: %s", err.Error())
+	}
+	defer osRemoveAll(tmpGPGDir)
+	for _, ppa := range classicStateMachine.ImageDef.Customization.ExtraPPAs {
+		ppaFileName, ppaFileContents := createPPAInfo(ppa,
+			classicStateMachine.ImageDef.Series)
+
+		ppaFile := filepath.Join(sourcesListD, ppaFileName)
+		ppaIO, err := osOpenFile(ppaFile, os.O_CREATE|os.O_WRONLY, 0644)
+		if err != nil {
+			return fmt.Errorf("Error creating %s: %s", ppaFile, err.Error())
+		}
+		ppaIO.Write([]byte(ppaFileContents))
+		ppaIO.Close()
+
+		// Import keys either from the specified fingerprint or via the Launchpad API
+		keyFileName := strings.Replace(ppaFileName, ".sources", ".gpg", 1)
+		keyFilePath := filepath.Join(classicStateMachine.tempDirs.chroot,
+			"etc", "apt", "trusted.gpg.d", keyFileName)
+		err = importPPAKeys(ppa, tmpGPGDir, keyFilePath)
+		if err != nil {
+			return fmt.Errorf("Error retrieving signing key for ppa \"%s\": %s",
+				ppa.PPAName, err.Error())
+		}
+	}
+	if err := osRemoveAll(tmpGPGDir); err != nil {
```

Comment:
Noob golang question: since we're already doing this via the `defer` functionality, do we need to explicitly do it here as well? Is it for proper error checking?